### ISSUE TITLE
Warning when loading data linked to plugin

### ISSUE THREAD

**Issue by Dobrodzieju** [veusz/veusz#493]
During season holidays I created several plugins which work well and prove that the concept of user plugins is very useful indeed.  I have small issue with one of them, which is related to spectral analysis. Veusz does not signal errors/warnings on start-up and during operation, when new data (power spectra) are created. However, when figure is saved with a new data linked to plugin and then, loaded back again , the following error shows up in the console window:

> C:\Program Files\Veusz\veusz\plugins\datasetplugin.py:54: ComplexWarning: Casting complex values to real discards the imaginary part

It is not harmful, because plugin works correctly as before, without new warning messages, but I wonder what maybe the reason for this. Below I attach the source code (plugin-spectral.py).
All the best in New Year 2021!

[plugin-spectral.zip](https://github.com/veusz/veusz/files/5758900/plugin-spectral.zip)

**Comment by jeremysanders** [comment]
Happy New Year!

Veusz tries to do a cast on the data using numpy.array(x, dtype=numpy.float64). If your code is returning a complex array then you'll get this warning.

I would suggest you put a print statement to print out the contents of the arrays before you pass them to dataset.update(). I think one of them has complex values. Alternatively, print `youarray.dtype` to see what data type them have.

**Comment by Dobrodzieju** [comment]
Thanks for the rapid response. My code obviously generates complex vector via
` y = fft.rfft(v*wind)`
and warning was apparently generated by (mathematically correct) line
`psd = fnorm * y * conjugate(y)`.
When it was replaced by the equivalent
`psd = fnorm * (y.real**2 + y.imag**2)`,
warning disappeared.
BTW, is it possible to make a new dataset, with `plugins.Dataset1D(..)`, which is _unlinked_ to plugin class by default?

**Comment by jeremysanders** [comment]
I don't think it can be unlinked, as the mechanism for updating the dataset relies on it being linked.

However, you could write a ToolsPlugin to do what you want without linking. They can do arbitrary things to the document, including getting datasets and setting datasets (e.g. with `SetData('dataset', numpy_array)`.
